aten/src/ATen/native/LossNLL.cpp

```diff
@@ -497,23 +497,78 @@ Tensor cross_entropy_loss_prob_target(
   }
 }
 
+Tensor cross_entropy_loss_label_smoothing(
+    const Tensor& self,
+    const Tensor& target,
+    const Tensor& weight,
+    int64_t reduction,
+    double label_smoothing) {
+
+    auto input = at::log_softmax(self, 1, self.scalar_type());
+    auto nllloss = at::nll_loss_nd(input, target, weight, Reduction::None);
+
+    auto n_classes = input.size(1);
+
+    auto eps_i = label_smoothing / (n_classes - 1);
+    Tensor smooth_loss;
+
+    if (weight.defined()) {
+      // Expand weight to the correct number of dims for broadcasting with input / target
+      auto weight_broadcast_shape = SmallBuffer<int64_t, 5>(input.dim());
+      std::fill(weight_broadcast_shape.begin(), weight_broadcast_shape.end(), 1);
+      weight_broadcast_shape[1] = weight.size(0);
+      Tensor weight_ = weight.view(weight_broadcast_shape);
+
+      smooth_loss = -at::sum(input * weight_, 1);
```

Comment:
Sorry if I misled you during our convo, but I think weight handling for the class index target case should stay consistent with what it's doing now. I could be wrong, but I think it'd be unexpected that setting label smoothing changes how weights are applied.